File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.js
function solution(topping) {
    let result = 0
    let [lCnt, rCnt] = [1, 0]
    let [lArr, rArr] = [Array(topping.length).fill(0), Array(topping.length).fill(0)]
	lArr[topping[0]] += 1
    for (let i = 1; i < topping.length; i += 1) {
        if (!rArr[topping[i]]) rCnt += 1
        rArr[topping[i]] += 1
    }
    for (let i = 1; i < topping.length; i += 1) {
        if (!lArr[topping[i]]) {
            lArr[topping[i]] += 1
            lCnt += 1
        }
        if (!(rArr[topping[i]] -= 1)) rCnt -= 1
        if (lCnt === rCnt) result += 1
    }
    return result
}

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.py
from collections import Counter

def solution(topping):
    result = 0
    
    left_counts = Counter([topping[0]])
    right_counts = Counter(topping[1:])
    
    left_unique_count = len(left_counts)
    right_unique_count = len(right_counts)
    
    for i in range(1, len(topping)):
        current_topping = topping[i]
        
        right_counts[current_topping] -= 1
        
        if right_counts[current_topping] == 0:
            right_unique_count -= 1
            
        if left_counts[current_topping] == 0:
            left_unique_count += 1
        left_counts[current_topping] += 1
        
        if left_unique_count == right_unique_count:
            result += 1

    return result